src/server.rs

```diff
@@ -120,6 +158,33 @@ impl LanguageServer for LspServer {
         }
         store.insert(key, value);
     }
+    async fn did_change(
+        &self,
+        params: lsp::DidChangeTextDocumentParams,
+    ) -> () {
+        let key = params.text_document.uri;
+        let mut store = self.store.lock().unwrap();
+        if !store.contains_key(&key) {
+            error!(
+                "textDocument/didChange called on unknown file {}",
+                key
+            );
+            return;
+        }
+        for change in params.content_changes {
+            if let Some(range) = change.range {
+                let contents = store.get(&key).unwrap();
+                let new_contents = replace_string_in_range(
+                    contents.clone(),
+                    range,
+                    change.text,
+                );
+                store.insert(key.clone(), new_contents);
+            } else {
+                store.insert(key.clone(), change.text);
+            }
+        }
+    }
```

Comment:
Just so I orient myself correctly here, the "server" here is the LSP server, and the client would be an editor of some kind, like the browser or VS Code or something. `textDocument/didChange` is a message sent from the client to the server to indicate that the user made a change to the file/document etc.

Reply:
Exactly. The LSP server has to keep a track of the changes that are being made in the client, and keeps a copy of the file in its own state; in this case, it's just a `HashMap<lsp_types::Url, String>`. It uses this representation when it needs to parse/format/etc. When you type things into your client, your client sends those change events to the server (hopefully in batches) so the server can update its own representation.